### DISCUSSION TITLE
how to configure streams-memory-audiokit

### DISCUSSION THREAD

**Discussion by bool-shark** [pschatzmann/arduino-audio-tools#614] [Q&A] [answered]
Hi Phil. I didn't see an example for memory streams to the audiokit's earphone jack so I'm trying unsuccessfully to modify the streams-memory-i2s sketch.
see below.  Please let me know what needs to be edited/added. Thanks
`
/**
 * @file streams-memory_mp3_short-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 stream and output as i2s signal
 * @version 0.1
 * @date 2021-01-24
 * 
 * @copyright Copyright (c) 2021
 
 */

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioKit.h"
#include "zero.h"

MemoryStream mp3(zero_mp3, zero_mp3_len);
AudioKitStream kit; 
MP3DecoderHelix helix;
EncodedAudioStream out(&out, &helix); // output to decoder
StreamCopy copier(kit, mp3);    // copy in to i2s

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // begin processing
  //auto cfg = i2s.defaultConfig();
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sample_rate = 24000;
  cfg.channels = 1;
  kit.setVolume(50);
  kit.begin(cfg);
  //out.begin();
}

void loop(){
  copier.copy();
  
/* if (mp3.available()) {
    copier.copy();
  } else {
    helix.end(); // flush output
    auto info = out.decoder().audioInfo();
    LOGI("The audio rate from the mp3 file is %d", info.sample_rate);
    LOGI("The channels from the mp3 file is %d", info.channels);
    i2s.end();
    stop();
  }*/
}

`

**Comment by pschatzmann** [comment, marked as answer]
This does not make any sense: if you do not change the cfg.output_device you should get output on both the speaker and on the earphone!

If you add a kit.processActions() to the loop the system will detect if you plugged in the earphones and will deactivate the speakers automatically.

**Reply by bool-shark** [reply]
Maybe I wasn't clear.  I don't want I2S only earphone output. Could you please highlight what needs to be changed. Thanks

**Comment by pschatzmann** [comment]
Just define the output device!
I never used this, so you need to figure out yourself if you need to use AUDIO_HAL_DAC_OUTPUT_LINE1  or AUDIO_HAL_DAC_OUTPUT_LINE2

**Comment by bool-shark** [comment]
Not sure if this is what you meant but it didn't work. 
I tried a different sketch that I was able to edit and it works (streams-memory_raw-i2s_external_dac.ino)  but doesn't compile when I add  kit.setVolume(); . I'll post it in a second post. 

#define AUDIO_CODEC_DEFAULT_CONFIG(){                   \
        .adc_input  = AUDIO_HAL_ADC_INPUT_LINE1,        \
        //.dac_output = AUDIO_HAL_DAC_OUTPUT_ALL,         \
        .dac_output =AUDIO_HAL_DAC_OUTPUT_LINE1;
		.codec_mode = AUDIO_HAL_CODEC_MODE_BOTH,        \
        .i2s_iface = {                                  \
            .mode = AUDIO_HAL_MODE_SLAVE,               \
            .fmt = AUDIO_HAL_I2S_NORMAL,                \
            .samples = AUDIO_HAL_48K_SAMPLES,           \
            .bits = AUDIO_HAL_BIT_LENGTH_16BITS,        \
        },                                              \

**Comment by pschatzmann** [comment]
Definitly no!  just add cfg.output_device = AUDIO_HAL_DAC_OUTPUT_LINE1!
```

auto cfg = kit.defaultConfig(TX_MODE);
cfg.sample_rate = 24000;
cfg.channels = 1;
cfg.output_device = AUDIO_HAL_DAC_OUTPUT_LINE1
kit.begin(cfg);
```

**Comment by bool-shark** [comment]
`
/**
   @file streams-memory_raw-i2s_external_dac.ino
   @author Phil Schatzmann
   @brief Compile with Partition Scheme Hughe APP!
   @version 0.1
   @date 2021-01-24

   @copyright Copyright (c) 2021

*/
#include "AudioTools.h"
#include "rassbwoy.h"
#include "AudioLibs/AudioKit.h"

uint8_t channels = 1;
uint16_t sample_rate = 44100;


AudioKitStream kit;
MemoryStream music(rassbwoy_raw, rassbwoy_raw_len);
StreamCopyT<int16_t> copier(kit, music); // copies sound into i2s


void setup() {
  Serial.begin(115200);
  Serial.println();
  

  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sample_rate = sample_rate;
  cfg.channels = channels;
  cfg.bits_per_sample = 16;
  //copier.setVolume(98);
  //kit.setVolume(90);
  kit.begin(cfg);
  

}

void loop() {
copier.copy();
 
}
`


AudioLibs/AudioKit.h: In member function 'virtual void audio_tools::AudioKitStream::setAudioInfo(audio_tools::AudioBaseInfo)':

AudioLibs/AudioKit.h:275:11: error: 'class AudioKit' has no member named 'setSampleRate'
       kit.setSampleRate(cfg.toSampleRate());

AudioKit.h: In member function 'audio_tools::AudioActions::ActiveLogic audio_tools::AudioKitStream::getActionLogic(int)':

AudioLibs/AudioKit.h:582:16: error: 'PERIPH_ID_ADC_BTN' was not declared in this scope

**Reply by bool-shark** [reply]
Oh wait just realized I didn't compile with huge app. Will try and report

**Comment by bool-shark** [comment]
it (streams-memory_raw-i2s_external_dac.ino) compiles and works with Setvolume but raw files are pretty big so honestly I'd rather get the sketch in the original post working. I tried your suggestion above , something else is still wrong. Help please.

`
/**
 * @file streams-memory_mp3_short-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 stream and output as i2s signal
 * @version 0.1
 * @date 2021-01-24
 * 
 * @copyright Copyright (c) 2021
 
 */

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioKit.h"
#include "zero.h"

MemoryStream mp3(zero_mp3, zero_mp3_len);
AudioKitStream kit; 
MP3DecoderHelix helix;
EncodedAudioStream out(&out, &helix); // output to decoder
StreamCopy copier(kit, mp3);    // copy in to i2s



void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  


auto cfg = kit.defaultConfig(TX_MODE);
cfg.sample_rate = 24000;
cfg.channels = 1;
cfg.output_device = AUDIO_HAL_DAC_OUTPUT_LINE1;
kit.begin(cfg);
//out.begin();
}

void loop(){
  copier.copy();
}
`

**Comment by pschatzmann** [comment]
What do you mean with something is wrong ?
Of cause the program will be very big if you decide to store all the audio in program memory!

**Comment by bool-shark** [comment]
Entire code below. I've gotten it to compile once but my audiokit was not connected at the time. Since then haven't been able to get it to compile. Same error messages from my previous post:

AudioLibs/AudioKit.h:275:11: error: 'class AudioKit' has no member named 'setSampleRate'
kit.setSampleRate(cfg.toSampleRate());

AudioKit.h: In member function 'audio_tools::AudioActions::ActiveLogic audio_tools::AudioKitStream::getActionLogic(int)':

AudioLibs/AudioKit.h:582:16: error: 'PERIPH_ID_ADC_BTN' was not declared in this scope.


`
/**
 * @file streams-memory_mp3_short-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 stream and output as i2s signal
 * @version 0.1
 * @date 2021-01-24
 * 
 * @copyright Copyright (c) 2021
 
 */

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "zero.h"
#include "AudioLibs/AudioKit.h"

MemoryStream mp3(zero_mp3, zero_mp3_len);
I2SStream i2s;
AudioKitStream kit;  
MP3DecoderHelix helix;
EncodedAudioStream out(&i2s, &helix); // output to decoder
StreamCopy copier(out, mp3);    // copy in to i2s

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // begin processing
  //auto cfg = i2s.defaultConfig();
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sample_rate = 24000;
  cfg.channels = 1;
  cfg.output_device = AUDIO_HAL_DAC_OUTPUT_LINE1;
  i2s.begin(cfg);
  out.begin();
}

void loop(){
  copier.copy();
  
}
`

**Comment by bool-shark** [comment]
I got it to compile but all I hear is static through audiokit earphone jack.

[I] I2SConfig.h : 104 - rx/tx mode: TX_MODE
[I] I2SConfig.h : 105 - port_no: 0
[I] I2SConfig.h : 106 - is_master: Master
[I] I2SConfig.h : 107 - sample rate: 24000
[I] I2SConfig.h : 108 - bits per sample: 16
[I] I2SConfig.h : 109 - number of channels: 1
[I] I2SConfig.h : 110 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 112 - auto_clear:0
[I] I2SConfig.h : 119 - buffer_count:10
[I] I2SConfig.h : 120 - buffer_size:512
[I] I2SConfig.h : 124 - pin_bck: 14
[I] I2SConfig.h : 125 - pin_ws: 15
[I] I2SConfig.h : 126 - pin_data: 0
[I] I2SStream.h : 68 - virtual void audio_tools::I2SStream::setAudioInfo(audio_tools::AudioBaseInfo)
[I] AudioTypes.h : 56 - sample_rate: 24000
[I] AudioTypes.h : 57 - channels: 1
[I] AudioTypes.h : 58 - bits_per_sample: 16
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 224 -> 224 -> 224 bytes - in 1 hops

**Comment by pschatzmann** [comment]
I could not reproduce your issue. However, I noticed that the current logic is overwriting the setting of output_device and was giving an output to both the speaker and the earphones, whatever settings were used.
I committed the correction to the [AudioKit](https://github.com/pschatzmann/arduino-audiokit) project

**Comment by bool-shark** [comment]
Are you saying the code exactly as I posted it my last post compiles with no errors and gives audio.
If you correct the errors of cause: Delete the I2SStream if you don't use it - this will point you to your mistakes
I was using https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/streams-memory_mp3-audiokit and just added the output_device!

**Comment by pschatzmann** [comment]
Yes, it has always been giving output to the earphones

**Reply by bool-shark** [reply]
I Got it,  thanks for your time, much appreciated.  I want to include 3 of these sound samples, 
do you have any ideas for how I could call them at different times.

**Reply by pschatzmann** [reply]
Look at the generated documentation for StreamCopy. You can dynamically define the source and destination with begin

**Reply by bool-shark** [reply]
hmm...ok will try thanks.
